marklogic-client-api/src/main/java/com/marklogic/client/datamovement/JSONSplitter.java

```diff
@@ -0,0 +1,550 @@
+                                    if (handle != null) {
+                                        return handle;
+                                    }
+                                    break;
+
+                                case SKIP:
+                                    jsonParser.skipChildren();
+                                    break;
+
+                                default:
+                                    throw new IllegalStateException("Unknown state");
+                            }
+
+                            break;
+
+                        case END_ARRAY:
+                            visitor.endArray(key.peek());
+                            break;
+                    }
+                }
+
+                return null;
+
+            } catch (IOException e) {
+                throw new RuntimeException("Failed to traverse document", e);
+            }
+        }
+
+    }
+
+    private static class HandleSpliterator<T extends AbstractWriteHandle> extends JSONSpliterator<T, T> {
+
+        HandleSpliterator(JSONSplitter<T> splitter, JsonParser jsonParser) {
+            super(splitter, jsonParser);
+        }
+
+        @Override
+        public boolean tryAdvance(Consumer<? super T> action) {
+            T handle = (T) getNextHandle();
+            if (handle == null) {
+                return false;
+            }
+
+            action.accept(handle);
+            getSplitter().count++;
+
+            return true;
+        }
+    }
+
+    private static class DocumentWriteOperationSpliterator<T extends AbstractWriteHandle> extends JSONSpliterator<DocumentWriteOperation, T> {
+
+        DocumentWriteOperationSpliterator(JSONSplitter<T> splitter, JsonParser jsonParser) {
+            super(splitter, jsonParser);
+        }
+
+        @Override
+        public boolean tryAdvance(Consumer<? super DocumentWriteOperation> action) {
+            T handle = (T) getNextHandle();
+            if (handle == null) {
+                return false;
+            }
+
+            DocumentWriteOperation documentWriteOperation = getSplitter().getVisitor().makeDocumentWriteOperation(handle);
+
+            action.accept(documentWriteOperation);
+            getSplitter().count++;
+
+            return true;
+        }
+    }
+
+    private static class JsonContainerParser extends JsonParserDelegate {
+
+        JsonContainerParser(JsonParser jsonParser) {
+            super(jsonParser);
+        }
+
+        private int depth = 1;
+
+        @Override
+        public boolean isClosed() {
+            return (depth > 0);
+        }
+
+        private void maintainDepth(JsonToken currentToken) {
+            if (currentToken == JsonToken.START_ARRAY || currentToken == JsonToken.START_OBJECT) {
+                depth++;
+            }
+
+            if (currentToken == JsonToken.END_ARRAY || currentToken == JsonToken.END_OBJECT) {
+                depth--;
+            }
```

Comment:
Would a switch be easier to understand?  The two cases that should be treated the same way can just fall through, as in:

```
switch(currentToken) {
    case START_ARRAY:
    case START_OBJECT:
        depth++;
        break;
...
} 
```

Reply:
Yes, changed the if/else clause to switch clause.